File: portal-frontend/src/utils/messageStatus.ts
// Message Status vocabulary — frontend twin of the backend module
// internal/shared/messagestatus. Single owner of the status enum, labels,
// badge variants and terminality; per-page label/color maps are banned.
//
// Domain rule (CONTEXT.md): `unknown` is never a stored status, so it has no
// entry here; unrecognized statuses fall back to a neutral chip via
// messageStatusMeta.

export const MESSAGE_STATUSES = [
  'pending',
  'queued',
  'sent',
  'delivered',
  'failed',
  'expired',
  'rejected',
  'scheduled',
  'cancelled',
] as const;

export type MessageStatus = (typeof MESSAGE_STATUSES)[number];

export type BadgeVariant = 'default' | 'success' | 'warning' | 'danger' | 'info';

export interface MessageStatusMeta {
  label: string;
  badgeVariant: BadgeVariant;
  /** Tailwind chip classes for pages rendering custom status spans. */
  colorClass: string;
  terminal: boolean;
}

export const messageStatus: Record<MessageStatus, MessageStatusMeta> = {
  pending: {
    label: 'Ожидание',
    badgeVariant: 'warning',
    colorClass: 'bg-yellow-100 text-yellow-800',
    terminal: false,
  },
  queued: {
    label: 'В очереди',
    badgeVariant: 'warning',
    colorClass: 'bg-yellow-100 text-yellow-800',
    terminal: false,
  },
  sent: {
    label: 'Отправлено',
    badgeVariant: 'info',
    colorClass: 'bg-blue-100 text-blue-800',
    terminal: false,
  },
  delivered: {
    label: 'Доставлено',
    badgeVariant: 'success',
    colorClass: 'bg-green-100 text-green-800',
    terminal: true,
  },
  failed: {
    label: 'Ошибка',
    badgeVariant: 'danger',
    colorClass: 'bg-red-100 text-red-800',
    terminal: true,
  },
  expired: {
    label: 'Истекло',
    badgeVariant: 'danger',
    colorClass: 'bg-gray-100 text-gray-600',
    terminal: true,
  },
  rejected: {
    label: 'Отклонено',
    badgeVariant: 'danger',
    colorClass: 'bg-red-100 text-red-800',
    terminal: true,
  },
  scheduled: {
    label: 'Запланировано',
    badgeVariant: 'info',
    colorClass: 'bg-purple-100 text-purple-800',
    terminal: false,
  },
  cancelled: {
    label: 'Отменено',
    badgeVariant: 'default',
    colorClass: 'bg-gray-100 text-gray-600',
    terminal: true,
  },
};

export function isMessageStatus(status: string): status is MessageStatus {
  return (MESSAGE_STATUSES as readonly string[]).includes(status);
}

/** Terminal statuses: no further transition expected. */
export function isTerminal(status: string): boolean {
  return isMessageStatus(status) && messageStatus[status].terminal;
}

/** Failure outcomes: the terminal statuses that did not deliver. */
export function isFailureOutcome(status: string): boolean {
  return status === 'failed' || status === 'rejected' || status === 'expired';
}

/** Status metadata with a neutral fallback for unrecognized values. */
export function messageStatusMeta(status: string): MessageStatusMeta {
  if (isMessageStatus(status)) return messageStatus[status];
  return { label: status, badgeVariant: 'default', colorClass: 'bg-gray-100 text-gray-700', terminal: false };
}

File: internal/shared/messagestatus/messagestatus.go
// Package messagestatus is the single owner of the Message Status
// vocabulary (CONTEXT.md): the status enum, terminality, and the mappers
// between Message status and the wire vocabularies (DLR stat codes, SMSC
// receipt formatting).
//
// Domain rule enforced here: `unknown` is NEVER a stored status — a receipt
// with an unrecognized stat leaves the Message in `sent` until the DLR
// timeout expires it (see FromDLRStat's ok=false).
//
// Raw status string literals are banned outside this package in
// Message-domain code (scripts/check-message-status.sh, wired into CI).
// Use shared.MessageStatus* aliases or this package directly.
package messagestatus

// Status is the Message lifecycle:
// pending → queued → sent → delivered | failed | expired | rejected,
// plus scheduled and cancelled.
type Status string

const (
	Pending   Status = "pending"
	Queued    Status = "queued"
	Sent      Status = "sent"
	Delivered Status = "delivered"
	Failed    Status = "failed"
	Expired   Status = "expired"
	Rejected  Status = "rejected"
	Scheduled Status = "scheduled"
	Cancelled Status = "cancelled"
)

// All lists every valid status. Order follows the lifecycle.
var All = []Status{Pending, Queued, Sent, Delivered, Failed, Expired, Rejected, Scheduled, Cancelled}

// Valid reports whether s is a member of the vocabulary.
func Valid(s Status) bool {
	switch s {
	case Pending, Queued, Sent, Delivered, Failed, Expired, Rejected, Scheduled, Cancelled:
		return true
	}
	return false
}

// IsTerminal reports whether no further transition is expected: the four
// terminal outcomes of the send lifecycle plus cancelled. pending, queued,
// scheduled and sent are non-terminal.
func IsTerminal(s Status) bool {
	switch s {
	case Delivered, Failed, Expired, Rejected, Cancelled:
		return true
	}
	return false
}

// FromDLRStat maps a DLR receipt `stat` code to a Message status.
//
// ok=false means the receipt must NOT change the stored status: unknown
// stat codes leave the Message in `sent` until the DLR timeout expires it
// (`unknown` is never a stored status — CONTEXT.md, Message Status).
func FromDLRStat(stat string) (Status, bool) {
	switch stat {
	case "DELIVRD":
		return Delivered, true
	case "EXPIRED":
		return Expired, true
	case "REJECTD", "UNDELIV":
		return Failed, true
	default:
		return "", false
	}
}

// ToSMSC maps a Message status to the receipt representation the platform
// emits when formatting a DLR for downstream consumers: the SMPP `stat`
// code and the `dlvrd` flag. Non-terminal statuses format as UNKNOWN
// receipts (callers only format receipts for terminal statuses).
func ToSMSC(s Status) (stat, dlvrd string) {
	switch s {
	case Delivered:
		return "DELIVRD", "001"
	case Failed:
		return "UNDELIV", "000"
	case Expired:
		return "EXPIRED", "000"
	case Rejected:
		return "REJECTD", "000"
	default:
		return "UNKNOWN", "000"
	}
}
